// @yext/pages/dist/plugins/onUrlChange/yext.js
const API_BASE_PROD = "https://api.yext.com/v2/accounts/me/";
const API_BASE_SBX = "https://api-sandbox.yext.com/v2/accounts/me/";
function buildApiUrl(base, path, params) {
  const result = new URL(path, base);
  for (const k in params) {
    result.searchParams.append(k, params[k]);
  }
  return result.toString();
}
async function updateEntity(id, locale, body, apiKey, options) {
  const URL_BASE = options?.env === "sbx" ? API_BASE_SBX : API_BASE_PROD;
  const url = buildApiUrl(URL_BASE, `entityprofiles/${id}/${locale}`, {
    api_key: apiKey,
    v: options?.v || "20220903"
  });
  const req = new Request(url, {
    method: "PUT",
    body: JSON.stringify(body),
    headers: { "Content-Type": "application/json; charset=utf-8" }
  });
  const response = await fetch(req);
  if (response.status < 200 || response.status >= 300) {
    const responseBody2 = await response.json();
    throw responseBody2.meta.errors[0].message;
  }
  const responseBody = await response.json();
  return responseBody.response;
}
export {
  buildApiUrl,
  updateEntity
};
