// @yext/pages/dist/dev/server/ssr/getLocalData.js
import path from "path";
import fs from "fs";
import { readdir } from "fs/promises";
import { findTemplateModuleInternal } from "./findTemplateModuleInternal.js";
import { validateGetPathValue } from "../../../common/src/template/internal/validateGetPathValue.js";
import { logWarning } from "../../../util/logError.js";
const LOCAL_DATA_PATH = "localData";
const getLocalDataManifest = async (vite, templateFilepaths) => {
  const localDataManifest = {
    static: /* @__PURE__ */ new Map(),
    entity: /* @__PURE__ */ new Map()
  };
  let dir;
  try {
    dir = await readdir(LOCAL_DATA_PATH);
  } catch (err) {
    if (err.code === "ENOENT") {
      return localDataManifest;
    } else {
      throw err;
    }
  }
  for (const fileName of dir) {
    const data = JSON.parse(
      fs.readFileSync(
        path.resolve(process.cwd(), `${LOCAL_DATA_PATH}/${fileName}`)
      ).toString()
    );
    const featureName = data.__?.name?.toString();
    if (!featureName) {
      continue;
    }
    const uid = data.uid?.toString();
    const entityId = data.id?.toString();
    const slug = data.slug?.toString();
    if (entityId) {
      localDataManifest.entity.set(featureName, [
        ...localDataManifest.entity.get(featureName) || [],
        { uid, entityId, slug }
      ]);
    } else {
      const templateModuleInternal = await findTemplateModuleInternal(
        vite,
        async (t) => featureName === t.config.name,
        templateFilepaths
      );
      if (!templateModuleInternal) {
        logWarning(
          `Could not find a static template for feature "${featureName}", skipping.`
        );
        continue;
      }
      const staticPath = templateModuleInternal.getPath({ document: data });
      const currentManifestData = localDataManifest.static.get(featureName);
      if (currentManifestData) {
        const occupiedPaths = currentManifestData.pathToLocaleMap;
        if (occupiedPaths.has(staticPath)) {
          throw new Error(
            `Path "${staticPath}" is used by multiple static pages.  Check that the getPath() function in the template "${templateModuleInternal.templateName}" returns a unique path for each locale.`
          );
        }
        occupiedPaths.set(staticPath, data.meta.locale);
      } else {
        try {
          validateGetPathValue(staticPath, templateModuleInternal.path);
        } catch (e) {
          logWarning(`${e.message}, skipping."`);
          continue;
        }
        const pathToLocaleMap = /* @__PURE__ */ new Map();
        pathToLocaleMap.set(staticPath, data.meta.locale);
        localDataManifest.static.set(featureName, {
          featureName,
          pathToLocaleMap
        });
      }
    }
  }
  return localDataManifest;
};
const getLocalData = async (criterion) => {
  try {
    const dir = await readdir(LOCAL_DATA_PATH);
    for (const fileName of dir) {
      const data = JSON.parse(
        fs.readFileSync(
          path.resolve(process.cwd(), `${LOCAL_DATA_PATH}/${fileName}`)
        ).toString()
      );
      if (criterion(data)) {
        return data;
      }
    }
  } catch (err) {
    if (err.code === "ENOENT") {
      throw "No localData exists. Please run `yext pages generate-test-data`";
    } else {
      throw err;
    }
  }
};
const getLocalDataForEntityOrStaticPage = async ({
  locale,
  featureName,
  entityId
}) => {
  const localData = await getLocalData((data) => {
    const isStatic = entityId === "";
    const isMatchingEntity = !isStatic && entityId === data.id;
    const matchesNameAndLocale = data.locale === locale && data.__.name === featureName;
    return (isStatic || isMatchingEntity) && matchesNameAndLocale;
  });
  if (!localData) {
    throw new Error(
      `No localData files match entityId, featureName, and locale: ${entityId}, ${featureName}, ${locale}`
    );
  }
  return localData;
};
const getLocalDataForSlug = async ({
  locale,
  slug
}) => {
  const localDataForSlug = (await getAllLocalData()).filter((d) => d.slug === slug);
  if (localDataForSlug.length === 0) {
    throw new Error(
      `No localData files match slug and locale: ${slug} ${locale}`
    );
  } else if (localDataForSlug.length > 1) {
    throw new Error(
      `Multiple localData files match slug and locale: ${slug} ${locale}, expected only a single file`
    );
  }
  return localDataForSlug[0];
};
const getAllLocalData = async () => {
  try {
    const dir = await readdir(LOCAL_DATA_PATH);
    return dir.map((fileName) => {
      const data = JSON.parse(
        fs.readFileSync(
          path.resolve(process.cwd(), `${LOCAL_DATA_PATH}/${fileName}`)
        ).toString()
      );
      return data;
    });
  } catch (err) {
    if (err.code === "ENOENT") {
      throw "No localData exists. Please run `yext pages generate-test-data`";
    } else {
      throw err;
    }
  }
};
export {
  getLocalDataForEntityOrStaticPage,
  getLocalDataForSlug,
  getLocalDataManifest
};
