// @yext/pages/dist/common/src/template/internal/types.js
import path from "node:path";
import { validateTemplateModuleInternal } from "./validateTemplateModuleInternal.js";
const parse = (filepath, adjustForFingerprintedAsset) => {
  let base = filepath.split(path.sep)[filepath.split(path.sep).length - 1];
  const extension = base.slice(base.lastIndexOf("."));
  let name = base.slice(0, base.lastIndexOf("."));
  if (adjustForFingerprintedAsset) {
    base = base.split(extension)[0].slice(0, base.split(extension)[0].lastIndexOf(".")) + extension;
    name = name.slice(0, name.lastIndexOf("."));
  }
  return {
    base,
    // the root file with extension
    name
    // the root file without extension
  };
};
const convertTemplateModuleToTemplateModuleInternal = (templateFilepath, templateModule, adjustForFingerprintedAsset) => {
  const templatePath = parse(templateFilepath, adjustForFingerprintedAsset);
  const templateModuleInternal = {
    ...templateModule,
    config: convertTemplateConfigToTemplateConfigInternal(
      templatePath.name,
      templateModule.config
    ),
    path: templateFilepath,
    filename: templatePath.base,
    templateName: templatePath.name
  };
  validateTemplateModuleInternal(templateModuleInternal);
  return templateModuleInternal;
};
const convertTemplateConfigToTemplateConfigInternal = (templateName, templateConfig) => {
  return {
    name: templateConfig?.name ?? templateName,
    hydrate: templateConfig?.hydrate ?? true,
    ...templateConfig,
    stream: convertStreamToStreamInternal(templateConfig?.stream)
  };
};
const convertStreamToStreamInternal = (stream) => {
  if (!stream)
    return;
  if (stream.localization.locales && stream.localization.locales.length > 0) {
    return {
      ...stream,
      localization: { locales: stream.localization.locales, primary: false }
    };
  }
  return { ...stream, localization: { primary: true } };
};
export {
  convertTemplateConfigToTemplateConfigInternal,
  convertTemplateModuleToTemplateModuleInternal,
  parse
};
