// @yext/pages/dist/upgrade/migrateConfig.js
import path from "path";
import fs from "fs";
import yaml from "yaml";
import { ProjectStructure } from "../common/src/project/structure.js";
import { logErrorAndExit } from "../util/logError.js";
const readJsonSync = (jsonPath) => {
  if (!fs.existsSync(jsonPath)) {
    return null;
  }
  try {
    return JSON.parse(fs.readFileSync(jsonPath, "utf-8"));
  } catch (e) {
    console.error(`unable to parse ${jsonPath}: ${e.message}`);
    return null;
  }
};
const writeYamlSync = (configYamlPath, target, data) => {
  const yamlDoc = yaml.parseDocument(fs.readFileSync(configYamlPath, "utf-8"));
  yamlDoc.set(target, data);
  fs.writeFileSync(configYamlPath, yaml.stringify(yamlDoc));
};
const migrateCiJson = async (configYamlPath, ciPath) => {
  const ciJson = readJsonSync(ciPath);
  if (ciJson !== null) {
    const buildArtifacts = ciJson.buildArtifacts;
    if (buildArtifacts) {
      console.info(
        `migrating buildArtifacts from ${ciPath} to ${configYamlPath}`
      );
      const buildConfiguration = {
        buildCommand: buildArtifacts.buildCmd?.replace("build:local", "build")
      };
      const dependencies = ciJson.dependencies;
      if (dependencies) {
        console.info(
          `migrating dependencies from ${ciPath} to ${configYamlPath}`
        );
        buildConfiguration.installDependenciesStep = {
          command: dependencies.installDepsCmd,
          requiredFiles: dependencies.requiredFiles
        };
      }
      writeYamlSync(configYamlPath, "buildConfiguration", buildConfiguration);
    }
    const livePreview = ciJson.livePreview;
    if (livePreview) {
      console.info(`migrating livePreview from ${ciPath} to ${configYamlPath}`);
      const livePreviewConfiguration = {
        setupCommand: livePreview.serveSetupCmd,
        serveCommand: livePreview.serveCmd,
        watchCommand: livePreview.watchCmd
      };
      writeYamlSync(
        configYamlPath,
        "livePreviewConfiguration",
        livePreviewConfiguration
      );
    }
  }
};
const migrateLocales = async (configYamlPath, featuresPath) => {
  const featuresJson = readJsonSync(featuresPath);
  if (!!featuresJson && !!featuresJson.locales) {
    console.info(`migrating locales from ${featuresPath} to ${configYamlPath}`);
    writeYamlSync(configYamlPath, "locales", featuresJson.locales);
  }
};
const migrateSiteStream = async (configYamlPath, siteStreamPath) => {
  const sitesJson = readJsonSync(siteStreamPath);
  if (sitesJson !== null) {
    console.info(
      `migrating global data from ${siteStreamPath} to ${configYamlPath}`
    );
    const newSiteStream = formatSiteStream(sitesJson, siteStreamPath);
    writeYamlSync(configYamlPath, "siteStream", newSiteStream);
  }
};
const formatSiteStream = (sitesJson, siteStreamPath) => {
  let entityId;
  if (sitesJson.filter?.entityIds && sitesJson.filter?.entityIds.length === 1) {
    entityId = sitesJson.filter.entityIds[0];
  } else if (sitesJson.filter?.entityIds) {
    logErrorAndExit(
      `Unable to migrate ${siteStreamPath} due to multiple entityIds`
    );
  }
  const siteStream = {
    id: sitesJson.$id,
    ...sitesJson,
    entityId
  };
  if (siteStream.reverseProxy?.displayUrlPrefix) {
    siteStream.serving = {
      reverseProxyPrefix: sitesJson.reverseProxy?.displayUrlPrefix
    };
  }
  delete siteStream.$id;
  delete siteStream.reverseProxy;
  delete siteStream.filter;
  return siteStream;
};
const migrateRedirects = async (source, dest) => {
  if (fs.existsSync(source)) {
    console.info(`migrating redirects from ${source} to ${dest}`);
    fs.copyFileSync(source, dest);
  }
};
const migrateServing = async (configYamlPath, servingPath) => {
  const servingJson = readJsonSync(servingPath);
  if (!!servingJson && !!servingJson.displayUrlPrefix) {
    console.info(
      `migrating reverse proxy info from ${servingPath} to ${configYamlPath}`
    );
    writeYamlSync(configYamlPath, "reverseProxy", {
      displayUrlPrefix: servingJson.displayUrlPrefix
    });
  }
};
const migrateSiteMap = async (configYamlPath, sitemapPath) => {
  const sitemapJson = readJsonSync(sitemapPath);
  if (sitemapJson !== null) {
    console.info(`migrating site map from ${sitemapPath} to ${configYamlPath}`);
    writeYamlSync(configYamlPath, "sitemap", sitemapJson);
  }
};
const migrateAuth = async (configYamlPath, authPath) => {
  const authJson = readJsonSync(authPath);
  if (authJson !== null) {
    console.info(`migrating auth info from ${authPath} to ${configYamlPath}`);
    writeYamlSync(configYamlPath, "authorization", authJson);
  }
};
const migrateConfigs = async (projectStructure) => {
  const scope = projectStructure.config.scope || "";
  const sitesConfigPath = projectStructure.getSitesConfigPath().getAbsolutePath();
  if (!fs.existsSync(sitesConfigPath)) {
    console.info("sites-config folder not found, nothing to migrate");
    return;
  }
  const files = fs.readdirSync(sitesConfigPath, { recursive: false });
  for (const file of files) {
    const filePath = path.resolve(sitesConfigPath, file.toString());
    if (fs.existsSync(filePath) && fs.statSync(filePath).isDirectory()) {
      await migrateConfigs(
        await ProjectStructure.init({
          scope: path.join(scope, file.toString())
        })
      );
    }
  }
  const sitesConfigFiles = projectStructure.config.sitesConfigFiles;
  const scopeFolder = path.resolve(scope);
  if (scope) {
    if (!fs.existsSync(scopeFolder)) {
      console.log(`creating scope folder for config at ${scopeFolder}`);
      fs.mkdirSync(scopeFolder, { recursive: true });
    }
  }
  const configYamlPath = path.resolve(
    scopeFolder,
    projectStructure.config.rootFiles.config
  );
  if (!fs.existsSync(configYamlPath)) {
    console.info(`${configYamlPath} does not exist, creating it`);
    fs.writeFileSync(configYamlPath, "");
  }
  await migrateCiJson(
    configYamlPath,
    path.resolve(sitesConfigPath, sitesConfigFiles.ci)
  );
  await migrateLocales(
    configYamlPath,
    path.resolve(sitesConfigPath, sitesConfigFiles.features)
  );
  await migrateSiteStream(
    configYamlPath,
    path.resolve(sitesConfigPath, sitesConfigFiles.siteStream)
  );
  await migrateServing(
    configYamlPath,
    path.resolve(sitesConfigPath, sitesConfigFiles.serving)
  );
  await migrateRedirects(
    path.resolve(sitesConfigPath, sitesConfigFiles.redirects),
    path.resolve(scopeFolder, sitesConfigFiles.redirects)
  );
  await migrateSiteMap(
    configYamlPath,
    path.resolve(sitesConfigPath, sitesConfigFiles.sitemap)
  );
  await migrateAuth(
    configYamlPath,
    path.resolve(sitesConfigPath, sitesConfigFiles.auth)
  );
  fs.rmSync(sitesConfigPath, { force: true, recursive: true });
};
export {
  formatSiteStream,
  migrateConfigs,
  readJsonSync
};
