// @yext/pages/dist/util/runtime.js
import { isNode, isDeno, isBrowser } from "browser-or-node";
class Runtime {
  name;
  /**
   * Whether or not the current runtime is being executed server-side or client-side. If the runtime
   * is node or deno then isServerSide will be true. When the runtime is browser isServerSide is
   * false.
   */
  isServerSide;
  version;
  constructor() {
    if (isDeno) {
      this.name = "deno";
      this.version = Deno?.version.deno || "";
      this.isServerSide = true;
    } else if (isNode) {
      this.name = "node";
      this.version = process.versions.node;
      this.isServerSide = true;
    } else if (isBrowser) {
      this.name = "browser";
      this.version = navigator.userAgent;
      this.isServerSide = false;
    } else {
      this.name = "unknown";
      this.version = "";
      this.isServerSide = false;
    }
  }
  getNodeMajorVersion() {
    if (this.name != "node") {
      throw new Error("Not running in Node.");
    }
    return +this.version.split(".")[0];
  }
}
const getRuntime = () => {
  return new Runtime();
};
export {
  getRuntime
};
